shell/platform/common/cpp/text_input_model.cc

```diff
@@ -39,26 +39,91 @@ void TextInputModel::SetText(const std::string& text) {
       utf16_converter;
   text_ = utf16_converter.from_bytes(text);
   selection_ = TextRange(0);
+  composing_range_ = TextRange(0);
 }
 
 bool TextInputModel::SetSelection(size_t base, size_t extent) {
-  size_t max_pos = text_.length();
-  if (base > max_pos || extent > max_pos) {
+  auto range = TextRange(base, extent);
+  if (composing_ && !range.collapsed()) {
     return false;
   }
-  selection_ = TextRange(base, extent);
+  if (!editing_range().Contains(range)) {
+    return false;
+  }
+  selection_ = range;
+  return true;
+}
+
+bool TextInputModel::SetComposingRange(size_t base,
+                                       size_t extent,
+                                       size_t cursor_offset) {
+  auto range = TextRange(base, extent);
+  if (!composing_ || !text_range().Contains(range)) {
+    return false;
+  }
+  composing_range_ = range;
+  selection_ = TextRange(range.start() + cursor_offset);
   return true;
 }
 
+void TextInputModel::BeginComposing() {
+  composing_ = true;
+  composing_range_ = TextRange(selection_.start());
+}
+
+void TextInputModel::UpdateComposingText(const std::string& composing_text) {
+  std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t>
+      utf16_converter;
+  std::u16string text = utf16_converter.from_bytes(composing_text);
+
+  // Preserve selection if we get a no-op update to the composing region.
+  if (text.length() == 0 && composing_range_.collapsed()) {
+    return;
+  }
+  DeleteSelected();
+  text_.replace(composing_range_.start(), composing_range_.length(), text);
+  SetComposingLength(text.length());
+  selection_ = TextRange(composing_range_.end());
+}
+
+void TextInputModel::CommitComposing() {
+  // Preserve selection if no composing text was entered.
+  if (composing_range_.collapsed()) {
+    return;
+  }
+  composing_range_ = TextRange(composing_range_.end());
+  selection_ = composing_range_;
+}
+
+void TextInputModel::EndComposing() {
+  composing_ = false;
+  composing_range_ = TextRange(0);
+}
+
 bool TextInputModel::DeleteSelected() {
   if (selection_.collapsed()) {
     return false;
   }
-  text_.erase(selection_.start(), selection_.length());
-  selection_ = TextRange(selection_.start());
+  size_t start = selection_.start();
+  text_.erase(start, selection_.length());
+  selection_ = TextRange(start);
+  if (composing_) {
+    // This occurs only immediately after composing has begun with a selection.
+    composing_range_ = selection_;
+  }
   return true;
 }
 
+void TextInputModel::SetComposingLength(size_t length) {
+  if (composing_range_.reversed()) {
+    size_t extent = composing_range_.extent();
+    composing_range_ = TextRange(extent + length, extent);
+  } else {
+    size_t base = composing_range_.base();
+    composing_range_ = TextRange(base, base + length);
+  }
+}
```

Comment:
We could avoid this by either (a) allowing TextRange to be mutable or (b) Adding a factory method to TextRange that returns a TextRange adjusted for length.

Reply:
I don't see any issues adding with more API surface to a class with such limited scope (and that's so easy to unit test). Mutable or mutate-copy both seem reasonable to me; whichever seems cleanest in practice to you SGTM.

Reply:
Done.